### PR TITLE
Allow registrants to opt out of announcing themselves when registering as a chat handler

### LINKED ISSUE #3358: Kafka Supervisor bloats zk discovery path
After testing kafka supervisor for two days, I got more than 1000 index_kafka_* entries in the zk discovery path and they don't go away (i.e. even stopping the supervisor does not help)

### PR DESCRIPTION
**Pull request by dclim**
Kafka indexing tasks were creating a large number of unnecessary announcements in ZK which caused performance issues, see https://groups.google.com/forum/#!topic/druid-development/wDEScqeQCp4

Fixes #3358

### CODE REVIEW

File: server/src/main/java/io/druid/segment/realtime/firehose/ServiceAnnouncingChatHandlerProvider.java [hunk 1]
```diff
@@ -52,24 +54,36 @@ public ServiceAnnouncingChatHandlerProvider(
...
-      serviceAnnouncer.announce(node);
-    }
-    catch (Exception e) {
-      log.warn(e, "Failed to register service[%s]", service);
-      handlers.remove(service, handler);
+    if (announce)
+    {
+      try {
+        serviceAnnouncer.announce(makeDruidNode(service));
+        if (!announcements.add(service)) {
```
Comment (fjy) [inline server/src/main/java/io/druid/segment/realtime/firehose/ServiceAnnouncingChatHandlerProvider.java]: do we need an ISE here or is a log.error enough?

File: server/src/main/java/io/druid/segment/realtime/firehose/ChatHandlerProvider.java [hunk 2]
```diff
@@ -25,9 +25,11 @@
  */
 public interface ChatHandlerProvider
 {
-  public void register(final String key, ChatHandler handler);
+  void register(final String key, ChatHandler handler);
 
-  public void unregister(final String key);
+  void register(final String key, ChatHandler handler, boolean announce);
```
Comment (fjy) [inline server/src/main/java/io/druid/segment/realtime/firehose/ChatHandlerProvider.java]: can we document this interface?

**Comment by fjy** [comment]
👍  just a few minor comments

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java [hunk 3]
```diff
@@ -534,6 +534,10 @@ public String apply(DataSegment input)
       log.info("The task was asked to stop before completing");
     }
 
+    if (chatHandlerProvider.isPresent()) {
+      chatHandlerProvider.get().unregister(getId());
+    }
```
Comment (nishantmonu51) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java]: should this be moved to finally block ?

File: server/src/main/java/io/druid/segment/realtime/firehose/ServiceAnnouncingChatHandlerProvider.java (hunk 1 above)
Comment (dclim) [inline server/src/main/java/io/druid/segment/realtime/firehose/ServiceAnnouncingChatHandlerProvider.java]: ISE probably makes more sense; if you had identical service names you could get this to happen, and if we just logged and carried on there's a chance you'd get things talking to the wrong things thinking that they were talking to other things.

File: server/src/main/java/io/druid/segment/realtime/firehose/ChatHandlerProvider.java (hunk 2 above)
Comment (dclim) [inline server/src/main/java/io/druid/segment/realtime/firehose/ChatHandlerProvider.java]: Sure

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java (hunk 3 above)
Comment (dclim) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java]: That sounds good, thanks!

**Comment by fjy** [comment]
👍

**Comment by nishantmonu51** [comment]
👍

**Pull request merged**